images/clustermesh-apiserver/Dockerfile

```diff
@@ -1,61 +1,66 @@
-#
-ARG TARGETARCH
-RUN make GOARCH=$TARGETARCH RACE=$RACE NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG
-WORKDIR /go/src/github.com/cilium/cilium
-RUN make GOARCH=$TARGETARCH licenses-all
 
-# CGO_ENABLED=0 GOOS=linux go build
+WORKDIR /go/src/github.com/cilium/cilium/clustermesh-apiserver
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+    mkdir -p /out/${TARGETOS}/${TARGETARCH} && cp etcd-config.yaml /out/${TARGETOS}/${TARGETARCH}/etcd-config.yaml
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+    make GOARCH=${TARGETARCH} RACE=${RACE} NOSTRIP=${NOSTRIP} LOCKDEBUG=${LOCKDEBUG} \
+    && mkdir -p /out/${TARGETOS}/${TARGETARCH}/usr/bin && mv clustermesh-apiserver /out/${TARGETOS}/${TARGETARCH}/usr/bin
+
+WORKDIR /go/src/github.com/cilium/cilium
+# licenses-all is a "script" that executes "go run" so it's ARCH should be set
+# to the same ARCH specified in the base image of this Docker stage
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+    make GOARCH=${BUILDARCH} licenses-all && mv LICENSE.all /out/${TARGETOS}/${TARGETARCH}
 
-FROM docker.io/library/alpine:3.13.1@sha256:08d6ca16c60fe7490c03d10dc339d9fd8ea67c6466dea8d558526b1330a85930 as certs
-ARG CILIUM_SHA=""
-LABEL cilium-sha=${CILIUM_SHA}
+# BUILDPLATFORM is an automatic platform ARG enabled by Docker BuildKit.
+# Represents the plataform where the build is happening, do not mix with
+# TARGETARCH
+FROM --platform=${BUILDPLATFORM} docker.io/library/alpine:3.13.1@sha256:08d6ca16c60fe7490c03d10dc339d9fd8ea67c6466dea8d558526b1330a85930 as certs
 RUN apk --update add ca-certificates
 
-# Cross-compile go, FROM comment must be located right before the FROM
-# line for the parameter to be applied on BuildKit builds.
-#
-# FROM --platform=$BUILDPLATFORM
-FROM docker.io/library/golang:1.15.8 as gops
-ARG CILIUM_SHA=""
-LABEL cilium-sha=${CILIUM_SHA}
-# TARGETARCH is an automatic platform ARG enabled by Docker BuildKit.
-#
-ARG TARGETARCH
-RUN GOARCH=$TARGETARCH && [ "$GOARCH" != "arm64" ] || CC="aarch64-linux-gnu-gcc" && \
-    go get -d github.com/google/gops && \
-    cd /go/src/github.com/google/gops && \
-    git checkout -b v0.3.14 v0.3.14 && \
-    git --no-pager remote -v && \
-    git --no-pager log -1 && \
-    CGO_ENABLED=0 go install && \
-    strip /go/bin/gops
+# BUILDPLATFORM is an automatic platform ARG enabled by Docker BuildKit.
+# Represents the plataform where the build is happening, do not mix with
+# TARGETARCH
+FROM --platform=${BUILDPLATFORM} docker.io/library/golang:1.15.8@sha256:9fdb74150f8d8b07ee4b65a4f00ca007e5ede5481fa06e9fd33710890a624331 as gops
+
+# build-gops.sh will build both archs at the same time
```

Comment:
The build steps are executed once for each target arch, so we are building gops four times (2x2)?